frame/staking/src/mock.rs

```diff
@@ -543,8 +605,29 @@ impl ExtBuilder {
 	}
 	pub fn build_and_execute(self, test: impl FnOnce() -> ()) {
 		sp_tracing::try_init_simple();
+		let check_events = self.check_events;
 		let mut ext = self.build();
+		ext.execute_with(|| {
+			// Clean up all the events produced on init.
+			OnStakeUpdate::take();
```

Comment:
All of this can be one `::take()` if we bundle them as one `enum`.

Reply:
sounds good!

Reply:
It's a little bit less convenient in terms using for assertions, as we are currently using `take` for that too, but that should be easy to resolve.

Reply:
Thinking about it more - it could be interesting also in terms of testing the sequence of different events that are being emitted, because with the original approach we could only see the sequence of the same type of event. 
It's going to make things a little bit more verbose in assertions, but with tests tailored specifically for those scenarios this should not be a problem.

Reply:
pattern matching on slices might help a lot here: 

```
assert!(matches!(&OnStakeUpadte::take()[..], [Event1, Event2, .. EventLast]))
```

etc.

Reply:
Yeah, that's the idea. It just gets a little bit verbose with too many events